### PR TITLE
Update C++ Autonomous Functionality

### PR DESCRIPTION
**Pull request by r4stered**
This updates the C++ version of the autonomous functionality to compile and be useable. It is different from the Java version because of CommandPtr being move only.  

I still need to add the AutoChooser class, but otherwise I have tested functionality here: https://github.com/r4stered/choreobetarobot

### CODE REVIEW

File: choreolib/src/main/native/include/choreo/Choreo.h [hunk 1]
```diff
@@ -59,7 +59,7 @@ class Choreo {
       std::error_code ec;
       auto fileBuffer =
           wpi::MemoryBuffer::GetFile(matchingFiles[0].string(), ec);
-      if (!fileBuffer || !ec) {
+      if (!fileBuffer || ec) {
```
Comment (spacey-sooty) [inline choreolib/src/main/native/include/choreo/Choreo.h:62]: We'll have to update this for 2025

File: choreolib/src/main/native/include/choreo/Choreo.h [hunk 2]
```diff
@@ -104,7 +104,7 @@ class Choreo {
 
     std::error_code ec;
     auto fileBuffer = wpi::MemoryBuffer::GetFile(trajectoryFileName, ec);
-    if (!fileBuffer || !ec) {
+    if (!fileBuffer || ec) {
```
Comment (spacey-sooty) [inline choreolib/src/main/native/include/choreo/Choreo.h:107]: ditto

**Comment by oh-yes-0-fps** [comment]
I'm really not a big fan of how infective the `year` generic is. What is it getting you with it being a generic/template?
99% of users will never use anything other than the latest year. Does cpp support generic defaults?

**Comment by spacey-sooty** [comment]
C++ supports generic defaults which I think would be a good addition here

File: choreolib/src/main/native/include/choreo/auto/AutoChooser.h [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+ * updated every cycle by calling the Update function in your Robot periodic
+ * function.
+ *
+ * You can retrieve the auto routine CommandPtr that is currently
+ * selected by calling the GetSelectedAutoRoutine() function.
+ *
+ * @tparam SampleType The type of samples in the trajectory.
+ * @tparam Year The field year.
+ */
+template <choreo::TrajectorySample SampleType, int Year>
```
Comment (spacey-sooty) [inline choreolib/src/main/native/include/choreo/auto/AutoChooser.h:47]:
```suggestion
template <choreo::TrajectorySample SampleType, int Year = 2024>
```

File: choreolib/src/main/native/include/choreo/auto/AutoChooser.h [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+ * Once the AutoChooser is made you can add auto routines to it using the
+ * AddAutoRoutine() function . Unlike Sendable Chooser this chooser has to be
+ * updated every cycle by calling the Update function in your Robot periodic
+ * function.
+ *
+ * You can retrieve the auto routine CommandPtr that is currently
+ * selected by calling the GetSelectedAutoRoutine() function.
+ *
+ * @tparam SampleType The type of samples in the trajectory.
+ * @tparam Year The field year.
```
Comment (spacey-sooty) [inline choreolib/src/main/native/include/choreo/auto/AutoChooser.h:45]:
```suggestion
 * @tparam Year The field year. Defaults to the current year.
```

File: choreolib/src/main/native/include/choreo/auto/AutoLoop.h [hunk 5]
```diff
@@ -17,23 +18,66 @@ namespace choreo {
  * This loop should **not** be shared across multiple autonomous routines.
  *
  * @tparam SampleType The type of samples in the trajectory.
+ * @tparam The field year (default: the current year).
  */
-template <choreo::TrajectorySample SampleType>
+template <choreo::TrajectorySample SampleType, int Year>
```
Comment (spacey-sooty) [inline choreolib/src/main/native/include/choreo/auto/AutoLoop.h:23]:
```suggestion
template <choreo::TrajectorySample SampleType, int Year = 2024>
```

File: choreolib/src/main/native/include/choreo/auto/AutoTrajectory.h [hunk 6]
```diff
@@ -35,6 +62,23 @@ static constexpr units::meter_t DEFAULT_TOLERANCE = 3_in;
 template <choreo::TrajectorySample SampleType, int Year>
```
Comment (spacey-sooty) [inline choreolib/src/main/native/include/choreo/auto/AutoTrajectory.h:62]:
```suggestion
template <choreo::TrajectorySample SampleType, int Year = 2024>
```

Comment (spacey-sooty) [reply choreolib/src/main/native/include/choreo/auto/AutoTrajectory.h:62]: Doc comment above needs update

**Review by spacey-sooty — CHANGES_REQUESTED** [review]
Some docs already assumed a default even though we didn't have one

**Comment by r4stered** [comment]
> C++ supports generic defaults which I think would be a good addition here

There is now a default param to the current year using the util class.

File: choreolib/src/main/native/include/choreo/auto/AutoTrajectory.h [hunk 7]
```diff
@@ -16,25 +18,69 @@
...
+   */
+  bool hasTriggered = false;
+};
+
 /**
  * A class that represents a trajectory that can be used in an autonomous
  * routine and have triggers based off of it.
  *
  * @tparam SampleType The type of samples in the trajectory.
  * @tparam Year The field year.
```
Comment (spacey-sooty) [inline choreolib/src/main/native/include/choreo/auto/AutoTrajectory.h:62]:
```suggestion
 * @tparam Year The field year. Defaults to the current year.
```

**Review by spacey-sooty — DISMISSED** [review]

**Comment by spacey-sooty** [comment]
@r4stered docs still need updates

**Review by spacey-sooty — APPROVED** [review]

**Review by calcmogul — APPROVED** [review]

**Pull request merged**